### ISSUE TITLE
counterCache joining irrelevant tables to update counter

### ISSUE THREAD

**Issue by johannesnagl** [cakephp/cakephp#5374]
Tested with CakePHP 2.5.1 (should be the same on current master)

We're using a counterCache in a scenario, where the parent-model has a lot of belongsTo and hasOne relations. Since counterCache uses `updateAll()` internally, all these relations are joined when updating the counter. the [documentation](http://book.cakephp.org/2.0/en/models/saving-your-data.html#model-updateall-array-fields-mixed-conditions) of `updateAll()` confirms this behavior (at least for updateAll).

Unfortunately, this is a problem when you want to have a mySQL Master/Slave scenario. Using `statement` based binlog format results in many warnings of this kind in the MySQL `error.log`:

`[Warning] Unsafe statement written to the binary log using statement format since BINLOG_FORMAT = STATEMENT. Statements writing to a table with an auto-increment column after selecting from another table are unsafe because the order in which rows are retrieved determines what (if any) rows will be written. This order cannot be predicted and may differ on master and the slave.`

This seems to stem form the current `\Mysql::update()` implementation which can possible [add joins to the update statement](https://github.com/cakephp/cakephp/blob/ac6d5cc70f2932de5eee17638ac7cd6c264ca9b0/lib/Cake/Model/Datasource/Database/Mysql.php#L405) (e.g. when using `updateAll` or pure `update` with certain conditions) when Models relations are expressed via [belongsTo](http://book.cakephp.org/2.0/en/models/associations-linking-models-together.html#belongsto) or when using the [`counterCache`](http://book.cakephp.org/2.0/en/models/associations-linking-models-together.html#countercache-cache-your-count).

It can be evaded by using either [`row` or `mixed` binlog_format](http://dev.mysql.com/doc/refman/5.6/en/binary-log-setting.html) but there are mixed reactions available to that issue:
- row based can be an issue if you've to update many rows
- mixed seems to combine the best of both worlds, however it seems the internal heuristic, albeit documented, seems to change with MySQL releases and still could contain race conditions

[There's a good read about the uncertainty of row/mixed based binlog format](http://www.percona.com/blog/2014/02/12/8-common-but-deadly-mysql-operations-mistakes-and-how-to-avoid-them/) on the Percona blog by Bill Karwin.

TL;DR:
- Q1: How to deterministically avoid JOINS in update statements?
- Q2: Why is counterCache joining irrelevant tables? there should be no reason at all to do this

(there was a stackoverflow thread about [this topic](http://stackoverflow.com/questions/25075032/cakephp-countercache-joining-irrelevant-tables-to-update-counter). because this is related to CakePHP-core, the answers provided did not solve the problem at all)
